feat: Public collections

SuaYoo opened the pull request:
Resolves https://github.com/webrecorder/browsertrix/issues/1051

Notes on the PR itself:
- The frontend code for commits have already been reviewed. This PR is open mostly for internal QA testing the UI.
- We should probably create a merge commit instead of the usual squash merging, since this includes a lot of changes.
- To manually test, see instructions and screenshots in individual PRs:
  - https://github.com/webrecorder/browsertrix/pull/2172
  - https://github.com/webrecorder/browsertrix/pull/2209
  - https://github.com/webrecorder/browsertrix/pull/2256

tw4l committed:
Make changes to collections to support publicly listed collections (#2164)

Fixes #2158 

- Adds `Organization.listPublicCollections` field and API endpoint to
update it
- Replaces `Collection.isPublic` boolean with `Collection.access`
(values: `private`, `unlisted`, `public`) and add database migration
- Update frontend to use `Collection.access` instead of `isPublic`,
otherwise not changing current behavior

---------

Co-authored-by: sua yoo <sua@suayoo.com>

tw4l committed:
Add public API endpoint for public collections (#2174)

Fixes #1051 

If org with provided slug doesn't exist or no public collections exist
for that org, return same 404 response with a detail of
"public_profile_not_found" to prevent people from using public endpoint
to determine whether an org exists.

Endpoint is `GET /api/public-collections/<org-slug>` (no auth needed) to
avoid collisions with existing org and collection endpoints.

SuaYoo committed:
feat: Public org profile page (#2172)

- Enables creating a public org profile page with description and
website at `/profile/<org slug>`
- Updates current "Overview" page to be "Dashboard", found under
`/dashboard`
- Organizes org "General" settings tab by "General", "Profile", and
"Developer Tools"
- Adds sign up banner to log in page for consistent CTA banners
- Updates copy and docs to support changes
- Allows user to set collection to private, public, or unlisted
- Adds route for public collection page with basic page layout
- Refactors copy button to abstract clipboard functionality
---------

Co-authored-by: Henry Wilkinson <henry@wilkinson.graphics>
Co-authored-by: emma <hi@emma.cafe>

SuaYoo committed:
feat: Make collection public (#2208)

tw4l committed:
Backend work for public collections (#2198)

Fixes #2182 

This rather large PR adds the rest of what should be needed for public
collections work in the frontend.

New API endpoints include:

- Public collections endpoints: GET, streaming download
- Paginated list of URLs in collection with snapshot (page) info for
each
- Collection endpoint to set home URL
- Collection endpoint to upload thumbnail as stream
- DELETE endpoint to remove collection thumbnail

Changes to existing API endpoints include:

- Paginating public collection list results
- Several `pages` endpoints that previously only supported `/crawls/` in
their path, e.g. `/orgs/{oid}/crawls/all/pages/reAdd`, now support
`/uploads/` and `/all-crawls/` namespaces as well. This is necessitated
by adding pages for uploads to the database (see below). For
`/orgs/{oid}/namespace/all/pages/reAdd`, `crawls` or `uploads` will
serve as a filter to only affect crawls of that given type. Other
endpoints are more liberal at this point, and will perform the same
action regardless of the namespace used in the route (we'll likely want
to change this in a follow-up to be more consistent).
- `/orgs/{oid}/namespace/all/pages/reAdd` now kicks off a background job
rather than doing all of the computation in an asyncio task in the
backend container. The background job additionally updates collection
date ranges, page/size counts, and tags for each collection in the org
after pages have been (re)added.

Other big changes:

- New uploads will now have their pages read into the database!
Collection page counts now also include uploads
- A migration was added to start a background job for each org that will
add the pages for previously-uploaded WACZ files to the database and
update collections accordingly
- Adds a new `ImageFile` subclass of `BaseFile` for thumbnails that we
can use for other user-uploaded image files moving forward, with
separate output models for authenticated and public endpoints

SuaYoo committed:
feat: Collection thumbnails, start page, and public view updates (#2209)

- Allows user to choose collection replay home page and collection
thumbnail (resolves
https://github.com/webrecorder/browsertrix/issues/2182)
- Displays collection thumbnails on org dashboard and public page
- Enables downloading public collection (resolves
https://github.com/webrecorder/browsertrix/issues/2233)
- Adds caption as "Summary" to metadata dialog
- Moves description editor to "About" tab

---------

Co-authored-by: Emma Segal-Grossman <hi@emma.cafe>

tw4l committed:
Add missing os import

SuaYoo committed:
feat: Track collection events (#2256)

- Renames `inject_analytics` to `inject_extra` and updates docs
- Manually tracks page views to enable passing custom props
- Tracks copying collection share link and downloading a public
collection

---------

Co-authored-by: emma <hi@emma.cafe>

SuaYoo committed:
chore: Refactor page headers (#2282)

- Refactors all page headers to use new `pageHeader`
- Removes border under org name/title in the org dashboard

ikreymer commented on backend/btrixcloud/basecrawls.py:
```diff
@@ -93,16 +88,8 @@ def __init__(
-
         # renew when <25% of time remaining
-        self.expire_at_duration_seconds = int(self.presign_duration_seconds * 0.75)
+        self.expire_at_duration_seconds = int(PRESIGN_DURATION_SECONDS * 0.75)
```
Do we want to keep this as configurable? Was this just a merge error?

tw4l replied:
It still is configurable! Just getting imported from models now so that we define it once and then import where we need to use it instead of defining the same thing multiple places

ikreymer replied:
Ah ok, missed that somehow!

ikreymer commented on backend/btrixcloud/colls.py:
```diff
@@ -391,10 +581,228 @@ async def add_successful_crawl_to_collections(self, crawl_id: str, cid: UUID):
+        page_size: int = DEFAULT_PAGE_SIZE,
+        page: int = 1,
+    ) -> Tuple[List[PageUrlCount], int]:
+        """List all URLs in collection sorted desc by snapshot count"""
```
Do we mean list all pages in collection, right?

tw4l replied:
It's actually URLs, not pages. The "pages" are stored in our database are actually snapshots (URL + timestamp combo specific to a crawl), so we may have multiple pages/snapshots for the same URL in a collection

ikreymer replied:
Hm, maybe we should call these Page URLs instead of just URLs, as that may imply all resources, including non pages?

ikreymer commented on backend/btrixcloud/colls.py:
```diff
@@ -391,10 +581,228 @@ async def add_successful_crawl_to_collections(self, crawl_id: str, cid: UUID):
+                flush=True,
+            )
+            await self.storage_ops.delete_file_object(org, thumbnail_file)
+            raise HTTPException(status_code=400, detail="upload_failed")
```
Should we pass this on as a more descriptive error, like `{"error": max_thumbnail_size_exceeded, "max_size": THUMBNAIL_MAX_SIZE}`

tw4l committed:
Make more explicit error message for large thumbnails

ikreymer commented on backend/btrixcloud/colls.py:
```diff
@@ -762,7 +762,10 @@ async def stream_iter():
-            raise HTTPException(status_code=400, detail="upload_failed")
+            raise HTTPException(
+                status_code=400,
+                detail="Upload failed: maxiumum thumbnail size (2 MB) exceeded",
```
would be better to have this be an id, so frontend can localize

tw4l replied:
Yes, fair. I think we're already validating this in the frontend before submission but can write it so it's useful for API users as well.

tw4l committed:
Use id for thumbnail size error detail

Shrinks99 committed:
ui: Public Collections UI Nitpicks (#2287)

- Removes share link from the dialogue footer
- Removes stickied collection navigation, replaces with improved
viewport-based scaling!
- Adds a max-width for the collection description in the logged in view.
- Moves the markdown editor buttons to below the editor
- Controls are now In-line with how we handle dialogue options
elsewhere, fixes a minor responsive design issue.
- Minor copy changes

---------

Co-authored-by: emma <hi@emma.cafe>
Co-authored-by: sua yoo <sua@webrecorder.org>
